README.md
# Network Resource Vocabulary

This is a repository to version NRV -- network resource vocabulary.

`pip install nrv`

## Usage

```
import nrv

nrv.list_versions()

nrv.get_version('v1')
```

## Development

Create a PR to suggested new version, name it as new file, with `v$`, where `$` is a natural number.


CHANGES.md
CHANGES

2021-02-21

* Add v3 version:
* Swap the 300 with 400.
* Introduce 600, and move locations from 100 to 600.


tests/test_validate.py
import pytest

from nrv import validate

def test_none():
    result = None
    expect = None
    assert(result == expect)

def test_():
    versionless_batch_1 = [
        '0oo.li/Intent::v1/category',
        '0oo.li/Method::v1/method',
        '0oo.li/Project::v1/system:organization',
        '0oo.li/Result::v1/resource',
        '0oo.li/Operation::v1/request:order',
        '0oo.li/Tool::v1/method:equipment',
    ]
    versioned_batch_2 = [
        '0oo.li/Place::v1/location',
        '0oo.li/Event::v1/resource:record',
        '0oo.li/Comment::v1/resource:record',
        '0oo.li/Product::v1/resource:method',
    ]
    failure_batch_3 = [
        '0oo.li/Order::v2/request:order',
        '0oo.li/Trade::v3/resource:record',
        '0oo.li/Person::v2/system:person'
    ]
    triple_batch_4 = [
        '0oo.li/Trade::v1/resource:operation:contract',
    ]
    v2_test_5 = [
        '0oo.li/Location::v2/resource:location',
        '0oo.li/Location::v2/150',
    ]

    b1 = all([validate(name, include_top=True) for name in versionless_batch_1])
    b2 = all([validate(name, include_top=True) for name in versioned_batch_2])
    b3 = not all([validate(name, include_top=True) for name in failure_batch_3])
    b4 = all([validate(name, include_top=True) for name in triple_batch_4])
    b5 = all([validate(name) for name in v2_test_5])

    assert(b1 and b2 and b3 and b4 and b5)


nrv/v1.py
data = {
    'category': [
        'goal',
        'concept',
        'question'
    ],
    'system': [
        'project',
        'structure',
        'mechanism',
        'organization',
        'person',
        'process',
        'api'
    ],
    'method': [
        'idea',
        'principle',
        'function',
        'invention',
        'equipment',
        'product',
        'endpoint'
    ],
    'location': [
        'address',
        'account',
        'url'
    ],
    'request': [
        'CATEGORY',
        'call',
        'call:query',
        'order',
        'order:buy',
        'order:sell',
        'order:transfer',
        'task',
        'task:programming',
        'task:engineering',
        'task:manufacturing',
        'task:treating',
        'operation',
        'operation:machining',
        'operation:shipment',
        'operation:payment',
        'operation:contract',
    ],
    'resource': [
        'record',
        'document',
        'METHOD',
        'SYSTEM'
    ],
}


nrv/v2.py
data = {
    '*': {'en': 'v2, data alignment protocol', 'cn': 'v2, 数据对齐协议', 'lt': 'v2, duomenų apjungimo protokolas'},
    100: {
        '*': {'en': 'resource', 'cn': '资源', 'lt': 'resursas'},
        110: {'en': 'record', 'cn': '记录', 'lt': 'įrašas'},
        111: {'en': 'record:identifier', 'cn': '记录:识别码', 'lt': 'įrašas:identifikatorius'},
        120: {'en': 'event', 'cn': '事件', 'lt': 'įvykis'},
        130: {'en': 'document', 'cn': '文档', 'lt': 'dokumentas'},
        140: {'en': 'material', 'cn': '资料', 'lt': 'medžiaga'},
        150: {'en': 'location', 'cn': '位置', 'lt': 'vieta'},
        151: {'en': 'location:address', 'cn': '位置:地址', 'lt': 'vieta:adresas'},
        152: {'en': 'location:account', 'cn': '位置:账户', 'lt': 'vieta:sąskaita'},
        153: {'en': 'location:url', 'cn': '位置:网址', 'lt': 'vieta:url'},
        160: {'en': 'location:ledger', 'cn': '位置:分类帐', 'lt': 'vieta:knyga'},
        200: {'en': 'CATEGORY', 'cn': '分类', 'lt': 'KATEGORIJA'},
        300: {'en': 'SYSTEM', 'cn': '系统', 'lt': 'SISTEMA'},
        400: {'en': 'METHOD', 'cn': '方法', 'lt': 'METODAS'},
        500: {'en': 'OPERATION', 'cn': '操作', 'lt': 'OPERACIJA'},
    },
    200: {
        '*': {'en': 'category', 'cn': '分类', 'lt': 'kategorija'},
        210: {'en': 'goal', 'cn': '目标', 'lt': 'tikslas'},
        220: {'en': 'concept', 'cn': '概念', 'lt': 'sąvoka'},
        230: {'en': 'question', 'cn': '问题', 'lt': 'klausimas'},
        240: {'en': 'property', 'cn': '特征', 'lt': 'savybė'},
    },
    300: {
        '*': {'en': 'system', 'cn': '系统', 'lt': 'sistema'},
        310: {'en': 'structure', 'cn': '结构', 'lt': 'struktūra'},
        311: {'en': 'structure:prototype', 'cn': '结构:原型', 'lt': 'struktūra:prototipas'},
        320: {'en': 'mechanism', 'cn': '机制', 'lt': 'mechanizmas'},
        330: {'en': 'organization', 'cn': '组织', 'lt': 'organizacija'},
        340: {'en': 'process', 'cn': '过程', 'lt': 'procesas'},
        350: {'en': 'api', 'cn': 'api', 'lt': 'api'},
        360: {'en': 'project', 'cn': '项目', 'lt': 'projektas'},
        370: {'en': 'person', 'cn': '个人', 'lt': 'individas'},
        200: {'en': 'CATEGORY', 'cn': '分类', 'lt': 'KATEGORIJA'},
    },
    400: {
        '*': {'en': 'method', 'cn': '方法', 'lt': 'metodas'},
        410: {'en': 'idea', 'cn': '想法', 'lt': 'idėja'},
        420: {'en': 'principle', 'cn': '原理', 'lt': 'principas'},
        430: {'en': 'procedure', 'cn': '程序', 'lt': 'procedūra'},
        440: {'en': 'function', 'cn': '函数', 'lt': 'funkcija'},
        450: {'en': 'invention', 'cn': '发明', 'lt': 'išradimas'},
        460: {'en': 'endpoint', 'cn': '端点', 'lt': 'terminalas'},
        470: {'en': 'product', 'cn': '产品', 'lt': 'produktas'},
        480: {'en': 'equipment', 'cn': '设备', 'lt': 'įranga'},
        300: {'en': 'SYSTEM', 'cn': '系统', 'lt': 'SISTEMA'},
    },
    500: {
        '*': {'en': 'operation', 'cn': '操作', 'lt': 'operacija'},
        510: {'en': 'call', 'cn': '调用', 'lt': 'kvietimas'},
        511: {'en': 'call:query', 'cn': '调用:询问', 'lt': 'kvietimas:užklausa'},
        520: {'en': 'order', 'cn': '订单', 'lt': 'orderis'},
        521: {'en': 'order:buy', 'cn': '订单:买单', 'lt': 'orderis:pirkimo'},
        522: {'en': 'order:sell', 'cn': '订单:卖单', 'lt': 'orderis:pardavimo'},
        523: {'en': 'order:transfer', 'cn': '订单:转移', 'lt': 'orderis:persiuntos'},
        524: {'en': 'order:machining', 'cn': '订单:加工', 'lt': 'orderis:apdirbimo'},
        525: {'en': 'order:shipment', 'cn': '订单:发货', 'lt': 'orderis:siuntos'},
        526: {'en': 'order:payment', 'cn': '订单:付款', 'lt': 'orderis:apmokėjimo'},
        527: {'en': 'order:contract', 'cn': '订单:合约', 'lt': 'orderis:sutarties'},
        528: {'en': 'order:evaluation', 'cn': '订单:评价', 'lt': 'orderis:vertinimo'},
        529: {'en': 'order:message', 'cn': '订单:消息', 'lt': 'orderis:žinutė'},
        530: {'en': 'task', 'cn': '任务', 'lt': 'užduotis'},
        531: {'en': 'task:programming', 'cn': '任务:编程', 'lt': 'užduotis:programavimo'},
        532: {'en': 'task:engineering', 'cn': '任务:工程', 'lt': 'užduotis:inžinerijos'},
        533: {'en': 'task:manufacturing', 'cn': '任务:制造', 'lt': 'užduotis:gamybos'},
        534: {'en': 'task:treatment', 'cn': '任务:处理', 'lt': 'užduotis:tvarkymo'},
        535: {'en': 'task:art', 'cn': '任务:美术', 'lt': 'užduotis:meno'},
        200: {'en': 'CATEGORY', 'cn': '分类', 'lt': 'KATEGORIJA'},
    },
}


nrv/utils.py
import importlib
import pkgutil
import os

from boltons.iterutils import remap
from collections import defaultdict


def get_valid_codes(data):
    result = defaultdict(list)

    def visit(path, key, value):
        if isinstance(key, int):
            if key % 100:
                result[100* (key // 100)].append(key)
        return key, value

    remap(data, visit=visit)
    return dict(result)


def get_valid_names(data, lang='en'):
    result = defaultdict(dict)

    def visit(path, key, value):
        if isinstance(key, int):
            if key % 100:
                key_in_words = 100 * (key // 100)
                vals_in_langs = data[key_in_words]
                vals_in_words = [item.get(lang) for item in vals_in_langs.values()]
                result[vals_in_words[0]] = vals_in_words[1:]
        return key, value

    remap(data, visit=visit)
    return dict(result)

def list_versions():
    result = []
    pkg = importlib.import_module('nrv')
    pkgpath = os.path.dirname(pkg.__file__)
    for _, name, _ in pkgutil.iter_modules([pkgpath]):
        if name.startswith('v') and name[1:].isalnum():
            result.append(name)
    return result

def get_latest_version():
    latest = max([int(v[1:]) for v in list_versions()])
    version = 'v%s' % latest
    return version


## MAIN FUNCTIONS ##


def get_version(version=None, exclude_uppercase=True, include_top=True):
    ''' names '''

    if version is None:
        version = get_latest_version()

    v =  importlib.import_module('nrv.%s' % version)
    result = []

    if version == 'v1':
        v_names = v.data
    else:
        v_names = get_valid_names(v.data)

    for key in v_names:
        if include_top:
            result.append('::%s' % key)
        for val in v_names[key]:
            if not val.isupper():
                result.append('::%s:%s' % (key, val))
    return result


def get_version_codes(version=None, include_top=True):
    '''' codes '''

    if version is None:
        version = get_latest_version()

    v =  importlib.import_module('nrv.%s' % version)

    if version == 'v1':
        v_codes = []
    else:
        v_codes = get_valid_codes(v.data)


    result = []
    for key in v_codes:
        if include_top:
            result.append('::%s:' % key)
        for val in v_codes[key]:
            result.append('::%s' % val)

    return result


nrv/__init__.py
from .utils import (
    list_versions,
    get_latest_version,
    get_valid_codes,
    get_valid_names,

    get_version,        # Returns valid names for version.
    get_version_codes,  # Returns valid codes for version.
)


def validate(codename, include_top=False):
    'validate table name'

    version = None

    if '::v' in codename:
        tok = codename.split('::v', 1)[-1]
        num = tok.split('/', 1)[0]
        if '/' in tok and num.isalnum():
            version = 'v%s' % num
            codename = '::'.join(codename.split('::%s/' % version))

            if version not in list_versions():
                return False

            # include_top -- False, cause there would be very many False positives, for now.
            valid_names = get_version(version, include_top=include_top)
            valid_codes = get_version_codes(version, include_top=include_top)

            valid_codenames = valid_codes + valid_names

    return any([token in codename for token in valid_codenames])




nrv/v3.py
data = {
    '*': {'en': 'v3, data alignment protocol', 'cn': 'v3, 数据对齐协议', 'lt': 'v3, duomenų apjungimo protokolas'},
    100: {
        '*': {'en': 'resource', 'cn': '资源', 'lt': 'išteklius'},
        110: {'en': 'record', 'cn': '记录', 'lt': 'įrašas'},
        111: {'en': 'record:identifier', 'cn': '记录:识别码', 'lt': 'įrašas:identifikatorius'},
        120: {'en': 'event', 'cn': '事件', 'lt': 'įvykis'},
        130: {'en': 'document', 'cn': '文档', 'lt': 'dokumentas'},
        140: {'en': 'material', 'cn': '资料', 'lt': 'medžiaga'},
        200: {'en': 'CATEGORY', 'cn': '分类', 'lt': 'KATEGORIJA'},
        300: {'en': 'SYSTEM', 'cn': '系统', 'lt': 'SISTEMA'},
        400: {'en': 'METHOD', 'cn': '方法', 'lt': 'METODAS'},
        500: {'en': 'OPERATION', 'cn': '操作', 'lt': 'OPERACIJA'},
        600: {'en': 'LOCATION', 'cn': '位置', 'lt': 'VIETA'},
    },
    200: {
        '*': {'en': 'category', 'cn': '分类', 'lt': 'kategorija'},
        210: {'en': 'goal', 'cn': '目标', 'lt': 'tikslas'},
        220: {'en': 'concept', 'cn': '概念', 'lt': 'sąvoka'},
        230: {'en': 'question', 'cn': '问题', 'lt': 'klausimas'},
        240: {'en': 'property', 'cn': '特征', 'lt': 'savybė'},
    },
    300: {
        '*': {'en': 'method', 'cn': '方法', 'lt': 'metodas'},
        310: {'en': 'idea', 'cn': '想法', 'lt': 'idėja'},
        320: {'en': 'principle', 'cn': '原理', 'lt': 'principas'},
        330: {'en': 'procedure', 'cn': '程序', 'lt': 'procedūra'},
        340: {'en': 'function', 'cn': '函数', 'lt': 'funkcija'},
        350: {'en': 'invention', 'cn': '发明', 'lt': 'išradimas'},
        360: {'en': 'endpoint', 'cn': '端点', 'lt': 'terminalas'},
        370: {'en': 'product', 'cn': '产品', 'lt': 'produktas'},
        380: {'en': 'equipment', 'cn': '设备', 'lt': 'įranga'},
        400: {'en': 'SYSTEM', 'cn': '系统', 'lt': 'SISTEMA'},
    },
    400: {
        '*': {'en': 'system', 'cn': '系统', 'lt': 'sistema'},
        410: {'en': 'structure', 'cn': '结构', 'lt': 'struktūra'},
        411: {'en': 'structure:prototype', 'cn': '结构:原型', 'lt': 'struktūra:prototipas'},
        420: {'en': 'mechanism', 'cn': '机制', 'lt': 'mechanizmas'},
        430: {'en': 'organization', 'cn': '组织', 'lt': 'organizacija'},
        440: {'en': 'process', 'cn': '过程', 'lt': 'procesas'},
        450: {'en': 'api', 'cn': 'api', 'lt': 'api'},
        460: {'en': 'project', 'cn': '项目', 'lt': 'projektas'},
        470: {'en': 'person', 'cn': '个人', 'lt': 'individas'},
        200: {'en': 'CATEGORY', 'cn': '分类', 'lt': 'KATEGORIJA'},
    },
    500: {
        '*': {'en': 'operation', 'cn': '操作', 'lt': 'operacija'},
        510: {'en': 'call', 'cn': '调用', 'lt': 'kvietimas'},
        511: {'en': 'call:query', 'cn': '调用:询问', 'lt': 'kvietimas:užklausa'},
        520: {'en': 'order', 'cn': '订单', 'lt': 'orderis'},
        521: {'en': 'order:buy', 'cn': '订单:买单', 'lt': 'orderis:pirkimo'},
        522: {'en': 'order:sell', 'cn': '订单:卖单', 'lt': 'orderis:pardavimo'},
        523: {'en': 'order:transfer', 'cn': '订单:转移', 'lt': 'orderis:persiuntos'},
        524: {'en': 'order:machining', 'cn': '订单:加工', 'lt': 'orderis:apdirbimo'},
        525: {'en': 'order:shipment', 'cn': '订单:发货', 'lt': 'orderis:siuntos'},
        526: {'en': 'order:payment', 'cn': '订单:付款', 'lt': 'orderis:apmokėjimo'},
        527: {'en': 'order:contract', 'cn': '订单:合约', 'lt': 'orderis:sutarties'},
        528: {'en': 'order:evaluation', 'cn': '订单:评价', 'lt': 'orderis:vertinimo'},
        529: {'en': 'order:message', 'cn': '订单:消息', 'lt': 'orderis:žinutė'},
        530: {'en': 'task', 'cn': '任务', 'lt': 'užduotis'},
        531: {'en': 'task:programming', 'cn': '任务:编程', 'lt': 'užduotis:programavimo'},
        532: {'en': 'task:engineering', 'cn': '任务:工程', 'lt': 'užduotis:inžinerijos'},
        533: {'en': 'task:manufacturing', 'cn': '任务:制造', 'lt': 'užduotis:gamybos'},
        534: {'en': 'task:treatment', 'cn': '任务:处理', 'lt': 'užduotis:tvarkymo'},
        535: {'en': 'task:art', 'cn': '任务:美术', 'lt': 'užduotis:meno'},
        200: {'en': 'CATEGORY', 'cn': '分类', 'lt': 'KATEGORIJA'},
    },
    600: {
        '*': {'en': 'location', 'cn': '位置', 'lt': 'vieta'},
        610: {'en': 'address', 'cn': '地址', 'lt': 'adresas'},
        620: {'en': 'account', 'cn': '账户', 'lt': 'sąskaita'},
        630: {'en': 'url', 'cn': '网址', 'lt': 'url'},
        640: {'en': 'ledger', 'cn': '分类帐', 'lt': 'knyga'},
    }
}
